### PR TITLE
Implement ISymUnmanagedBinder4, ISymUnmanagedReader4

### LINKED ISSUE #6184: Microsoft.DiaSymReader.PortablePdb.SymBinder.GetReaderFile
Hi Roslyn team(@tmat),
For this Microsoft.DiaSymReader.PortablePdb.SymBinder.GetReaderFile api, based on the source code seems the searchPath is useless.  The only way to search the file is just based on the file name. I guess maybe we should look at how CCI implement this feature. Using searchPath to get the reader is one of the key scenario which shouldn't be missed.

Thanks,
Zhicheng Zhu

### LINKED ISSUE #6420: [Portable Pdb] Need 'CheckMatch' method on SymBinder
In the debugger's symbol provider, we have the data from a debug directory and we need to be able to test if this matches a given portable PDB, and if so obtain an ISymUnmanagedReader. We are open to the exact shape of this API, but here is something that would work well for us:

[PreserveSig]
int GetAndValidateReader(string pdbPath, GUID mvid, int age, ISymUnmanagedMetadataImportProvider importProvider, [out] ISymUnmanagedReader reader);

interface ISymUnmanagedMetadataImportProvider
{
   object GetMetadataImport();
}

### PR DESCRIPTION
**Pull request by tmat**
Implements ISymUnmanagedBinder4 and ISymUnmanagedReader4 interfaces,
fixes implementation of GetReaderForFile to use information from PE debug directory and search paths to locate the .pdb.

The feature addressed by the change is asked for in #6420. The implementation provides different APIs that proposed in #6420 but still addresses the scenario efficiently. The main goal of the new ISymUnmanagedBinder4 API is to be able to create SymReader w/o creating the corresponding IMetadataImport (important in remote debugging scenarios where we want to avoid downloading the PE file unless absolutely necessary). The implementation actually goes further than originally proposed. It defers creation of the metadata import until SymReader API that need it are called. Since only a couple of APIs need to query the metadata this change should avoid creating the import in most debugging scenarios.

Fixes #6420.
Fixes #6184.

### CODE REVIEW

**Comment by tmat** [comment]
@gregg-miskelly @ZhichengZhu @noahfalk
FYI @dotnet/roslyn-interactive

**Comment by tmat** [comment]
This change is up for review, test failures failures are expected. I need to split this change into two - first we need to update the DiaSymReader nuget and then the rest can be checked in.

File: src/Debugging/Microsoft.DiaSymReader.PortablePdb/SymBinder.cs [hunk 1]
```diff
@@ -11,106 +17,357 @@ namespace Microsoft.DiaSymReader.PortablePdb
...
+            // 4b) registry
+            if ((searchPolicy & SymUnmanagedSearchPolicy.AllowRegistryAccess) != 0)
             {
-                reader = null;
-                return HResult.E_INVALIDARG;
+                // TODO
+            }
+
+            // 5c) environment variables:
+            yield return PortableShim.Environment.GetEnvironmentVariable("_NT_ALT_SYMBOL_PATH");
```
Comment (gregg-miskelly) [inline src/Debugging/Microsoft.DiaSymReader.PortablePdb/SymBinder.cs]: Should you have an 'if windows' check for these? Do you only want to yield return them if they are non-null?

File: src/Debugging/Microsoft.DiaSymReader.PortablePdb/PortablePdbReader.cs [hunk 2]
```diff
@@ -52,7 +103,7 @@ public void Dispose()
             if (!IsDisposed)
             {
                 _pinnedImage.Free();
-                MetadataImport = null;
```
Comment (gregg-miskelly) [inline src/Debugging/Microsoft.DiaSymReader.PortablePdb/PortablePdbReader.cs]: The CLR's implementation of IMetaDataImport will leave the file locked until it is final released, which means that we unfortunately cannot just rely on the GC to collect this. So, on Windows at least, we need to do --
if (_metaDataReader.HasValue && Marshal.IsComObject(_metaDataReader.Value))
{
   Marshal.ReleaseComObject(_metaDataReader.Value);
}

File: src/Debugging/Microsoft.DiaSymReader.PortablePdb/SymBinder.cs (hunk 1 above)
Comment (tmat) [inline src/Debugging/Microsoft.DiaSymReader.PortablePdb/SymBinder.cs]: The caller is handling null values.

File: src/Debugging/Microsoft.DiaSymReader.PortablePdb/PortablePdbReader.cs (hunk 2 above)
Comment (tmat) [inline src/Debugging/Microsoft.DiaSymReader.PortablePdb/PortablePdbReader.cs]: :(
Good catch, will fix.

File: src/Dependencies/Microsoft.DiaSymReader/IMetadataImportProvider.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    [ComVisible(false)]
+    public interface IMetadataImportProvider
+    {
+        /// <summary>
+        /// Gets an instance of IMetadataImport.
+        /// </summary>
+        /// <remarks>
+        /// The implementer is responsible for managing the lifetime of the resulting object.
+        /// </remarks>
+        object GetMetadataImport();
```
Comment (gregg-miskelly) [inline src/Dependencies/Microsoft.DiaSymReader/IMetadataImportProvider.cs]: Do you know if we need --
[MarshalAs(UnmanagedType.Interface)]

Or will the CLR assume this?

**Comment by gregg-miskelly** [comment]
Looks good to me aside from those comments.

File: src/Dependencies/Microsoft.DiaSymReader/IMetadataImportProvider.cs (hunk 3 above)
Comment (tmat) [inline src/Dependencies/Microsoft.DiaSymReader/IMetadataImportProvider.cs]: will add.

**Comment by tmat** [comment]
Fixes #6184

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
test win please (git error)

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
test win please (git error)

**Comment by tmat** [comment]
Failure: #7643
retest win please

**Pull request merged**